libraries/chain/chain_controller.cpp

```diff
@@ -1074,6 +1081,11 @@ void chain_controller::_initialize_chain(contracts::chain_initializer& starter)
             transaction_metadata tmeta( genesis_setup_transaction );
             __apply_transaction( tmeta );
          });
+
+         // TODO: Should we write this genesis block instead of faking it on startup?
+         block.input_transactions.emplace_back(genesis_setup_transaction);
+         applied_irreversible_block(block);
```

Comment:
in order to help indexers be generic we should have a guarantee that every block has at least `applied_block` called on it and only those that become irreversible eventually have `applied_irreversible_block` called.  

The take away here is that we should be building a full false block complete with trace of transactions to pass to indexers in both callbacks (in the proper order).  This would also allow the construction of merkle proofs that includ genesis transactions.

Reply:
Added `block_trace` and call to `applied_block`. Also called `calculate_root()` on `shard_trace`. Anything missing?